cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,555 @@
+   * @return a new {@link Map} where the key is the parameter name given in the annotation and value is the
+   * position of the parameter.
+   */
+  private static Map<String, Integer> processParameterAnnotationNode(Map<Integer, AnnotationNode>
+                                                                       parameterAnnotationNode) {
+    Map<String, Integer> parameterAnnotation = new HashMap<>();
+    for (Map.Entry<Integer, AnnotationNode> entry : parameterAnnotationNode.entrySet()) {
+      // the AnnotationNode of parameter will be an array of size 2 where 0 is "value" and 1 is the name provided in
+      // the annotation.
+      String paraName = (String) entry.getValue().values.get(1);
+      // We expect all parameters to have unique names to ensure that here
+      Preconditions.checkArgument(!parameterAnnotation.containsKey(paraName),
+                                  String.format("A parameter with name %s was already found at position %s." +
+                                                  " Please use unique names.", paraName,
+                                                parameterAnnotation.get(paraName)));
+      // store the parameter position with the unique name given to it
+      parameterAnnotation.put(paraName, entry.getKey());
+    }
+    return parameterAnnotation;
+  }
+
+  /**
+   * A class which can process {@link AuthEnforce} {@link AnnotationNode}
+   */
+  private static class AuthEnforceAnnotationNodeProcessor {
+
+    final Set<Action> actions = new HashSet<>();
+    Type enforceOn = null;
+    List<String> entities = null;
+
+    AuthEnforceAnnotationNodeProcessor(AnnotationNode annotationNode) {
+      List values = annotationNode.values;
+      // AuthEnforce AnnotationNode will have 6 values: 3 field names and another 3 field containing their values.
+      Preconditions.checkArgument(values.size() == 6, String.format("AuthEnforce annotation has three field and " +
+                                                                      "with their values total elements should be 6 " +
+                                                                      "but found %s", values.size()));
+      for (int i = 0; i < 6; i = i + 2) { // increment by 2 to go to next field
+        Object name = values.get(i);
+        if ("entities".equals(name)) {
+          entities = (List<String>) values.get(i + 1);
+        } else if ("enforceOn".equals(name)) {
+          enforceOn = (Type) values.get(i + 1);
+        } else if ("actions".equals(name)) {
+          // Actions is a ArrayList<String[]> where each String array is of size 2. String[0] is class name Action
+          // class and String[1] is the action which was specified in the annotation
+          List<String[]> actionWithClassName = (List<String[]>) values.get(i + 1);
+          for (String[] action : actionWithClassName) {
+            Preconditions.checkArgument(action.length == 2);
```

Comment:
Is it necessary ? Since the annotation defines Action[] and passing anything else will be compile time failure. Also, all these methods are private so can't be invoked directly from other class with incorrect data.